static/js/Editor.jsx

```diff
@@ -2546,18 +2547,152 @@ const BlockButton = ({format, icon}) => {
+          const img = new Image();
+          img.src = src;
+          img.onload = () => loadedIcons.add(src);  // Mark as loaded after successful load
+        }
+      });
+    }, [state]);
+
+    const tooltip = stateToTooltip[state];
+
+    return (
+        <ToolTipped altText={localize(tooltip)} classes={`editorSaveStateIndicator tooltip-toggle ${state}`}>
+        {<img src={stateToIcon[state]} alt={localize(state)} />}
+        <span className="saveStateMessage">{localize(stateToMessage[state])}</span>
+        </ToolTipped>
+  );
+}
+function useUnsavedChangesWatcher(timeoutSeconds, unsavedChanges, savingState, setSavingState) {
+  const timeoutRef = useRef(null);
+
+  // if failed to save and no reason is given, transition into known error state after a timeout
+  useEffect(() => {
+    const hasUnsaved = unsavedChanges;
+    if (hasUnsaved && !timeoutRef.current) {
+      // Start a timeout only if none is running
+      timeoutRef.current = setTimeout(() => {
+        savingState === "saving" && setSavingState('unknownError');
+        timeoutRef.current = null; // Reset the ref
+      }, timeoutSeconds * 1000);
+    }
+
+    if ((!hasUnsaved)
+        && timeoutRef.current) {
+      // Cancel if unsavedChanges were cleared before timeout ends
+      clearTimeout(timeoutRef.current);
+      timeoutRef.current = null;
+    }
+
+    return () => {
+      // Clean up on unmount or re-run
+      if (timeoutRef.current) {
+        clearTimeout(timeoutRef.current);
+        timeoutRef.current = null;
+      }
+    };
+  }, [unsavedChanges, savingState]);
+}
+
 const SefariaEditor = (props) => {
     const editorContainer = useRef();
+    const editorContentContainer = useRef();
+    const editorTitleContainer = useRef();
     const [sheet, setSheet] = useState(props.data);
     const initValue = [{type: "sheet", children: [{text: ""}]}];
     const renderElement = useCallback(props => <Element {...props}/>, []);
     const [value, setValue] = useState(initValue);
     const [currentDocument, setCurrentDocument] = useState(initValue);
     const [unsavedChanges, setUnsavedChanges] = useState(false);
+    const [blockEditing, setBlockEditing] = useState(false);
     const [lastModified, setlastModified] = useState(props.data.dateModified);
     const [canUseDOM, setCanUseDOM] = useState(false);
     const [lastSelection, setLastSelection] = useState(null)
     const [readyForNormalize, setReadyForNormalize] = useState(false);
+    const [connectionLostPolling, setConnectionLostPolling] = useState(false);
+    const [userUnauthenticated, setUserUnauthenticated] = useState(false);
+    const [unknownErrorDetected, setUnknownErrorDetected] = useState(false);
+    const savingState = props.newEditorSaveState;
+    const setSavingState = props.setNewEditorSaveState;
+    const isMultiPanel = Sefaria.multiPanel;
+    useUnsavedChangesWatcher(20, unsavedChanges, savingState, setSavingState);
+
+    useEffect(() => {
```

Comment:
I see 4 new useEffects. It is much easier to understand how useEffects work when they are pulled out as standalone hooks. This is because you need to give a name to the hook and you can bundle related useEffects in one hook. Are any of these candidates for being standalone hooks? I feel like useEffect 2 and 3 can be combined into one hook.